lib/revision.rb

```diff
@@ -0,0 +1,106 @@
+class Metis
+  class Revision
+    FILENAME_MATCH=/[^<>:;,?"*\|\/\x00-\x1f]+/x
+
+    FILEPATH_MATCH=%r!
+      \A
+        metis:\/\/
+        (?<project_name>(#{FILENAME_MATCH.source}))
+        /
+        (?<bucket_name>(#{FILENAME_MATCH.source}))
+        /
+        (?<file_path>(#{FILENAME_MATCH.source}).*)
+      \z
+    !x
+
+    attr_reader :source, :dest
+    def initialize(params)
+      raise Etna::BadRequest, 'All revisions require "source" parameter' unless params.has_key? :source
+      raise Etna::BadRequest, 'All revisions require "dest" parameter' unless params.has_key? :dest
+
+      raise Etna::BadRequest, 'Invalid source path' unless valid_file_path?(params[:source])
```

Comment:
Since we can't be sure we're in a request here, we might not want to raise BadRequest to get out of here. Another option might be to just save the params and save these validations for a `valid?` method, which the controller can use to check if the request is valid and then raise BadRequest, etc. If not that we might raise out of here with some other error that we catch in the controller, but since this object has nothing better to do than provide validations it seems fine to leave them as methods rather than raising.

Reply:
I didn't pass in the request -- thoughts on this `valid?` method? For bulk copy, I was trying to avoid fetching bucket for each revision, so passing in the list of authorized bucket names directly.